tests/Knp/Menu/Tests/Matcher/Voter/CallbackVoterTest.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Knp\Menu\Tests\Matcher\Voter;
+
+use Knp\Menu\ItemInterface;
+use Knp\Menu\Matcher\Voter\CallbackVoter;
+use PHPUnit\Framework\TestCase;
+
+final class CallbackVoterTest extends TestCase
+{
+    public function testNoMatchCallbackSet(): void
+    {
+        $item = $this->getMockBuilder(ItemInterface::class)->getMock();
+        $item->expects($this->once())
+            ->method('getExtra')
+            ->with('match_callback')
+            ->willReturn(null);
+
+        $voter = new CallbackVoter();
+
+        $this->assertNull($voter->matchItem($item));
+    }
+
+    public function testMatchCallbackIsNotCallable(): void
+    {
+        $item = $this->getMockBuilder(ItemInterface::class)->getMock();
+        $item->expects($this->once())
+            ->method('getExtra')
+            ->with('match_callback')
+            ->willReturn('foo');
+
+        $voter = new CallbackVoter();
+
+        $this->expectException(\InvalidArgumentException::class);
+
+        $voter->matchItem($item);
+    }
+
+    /**
+     * @dataProvider provideData
+     */
+    public function testMatching(callable $callback, ?bool $expected): void
+    {
+        $item = $this->getMockBuilder(ItemInterface::class)->getMock();
+        $item->expects($this->once())
+            ->method('getExtra')
+            ->with('match_callback')
+            ->willReturn($callback);
+
+        $voter = new CallbackVoter();
+
+        $this->assertSame($expected, $voter->matchItem($item));
+    }
+
+    /**
+     * @return array<string, array<int, callable|bool|null>>
+     */
+    public function provideData(): array
+    {
+        return [
+            'matching' => [fn (): ?bool => true, true],

```

Comment:
I suggest yielding the various cases (using `iterable` instead of `array` as return type) instead of returning an array. It makes it easier to identify where each case starts and ends in the data provider as the `yield` keywords starts each case.